src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java

```diff
@@ -925,14 +935,18 @@ public long transferFrom(ReadableByteChannel src,
         ensureOpen();
         if (!src.isOpen())
             throw new ClosedChannelException();
+        if (src instanceof FileChannelImpl fci && !fci.readable)
+            throw new NonReadableChannelException();
         if (!writable)
             throw new NonWritableChannelException();
         if ((position < 0) || (count < 0))
             throw new IllegalArgumentException();
         if (position > size())
             return 0;
 
-        if (src instanceof FileChannelImpl fci) {
+        // System calls supporting fast transfers might not work on files
+        // which advertise zero size such as those in Linux /proc
+        if (src instanceof FileChannelImpl fci && fci.size() > 0) {
```

Comment:
This means an extra fstat per transferFrom but maybe you've measured it with small files and it's okay?

Reply:
It's about 20% slower for sizes 32 and 4096 (disk block size), and about 13% slower for size 32768. Instead making the direct and trusted methods fail independently does not help.